### PR TITLE
fix: backward compatibility with version 8.0.0

### PR DESCRIPTION
**Pull request by salvatore-campagna**
Commit `fb6e1fb127e3c2bb1683402e9ab083115a56b08d` did not land into version `8.0.0`. As a result, we need to make master (version `8.2.0`) interoperate with version `8.0.0`.
This means we need to do the following:
* always read and write a **non-optional string key** for a range bucket (not for the range itself but for the bucket). In that case the key is used when serialising the response back to clients.
* write `originalFrom` and `originalTo` values so that a node running version `8.0.0` can read them properly.
* read `originalFrom` and `originalTo` from a node running version `8.0.0` and using them to **overwrite** `from` and `to` when creating a new range bucket.

All of this needs to be done keeping compatibility with version `7.17` which was fixed by PR #83715 .
Also note that, interoperability with version `8.0.1` is taken care by PR #83715. As a result, here we really just need to address backward compatibility with version `8.0.0` that is already released and we cannot change.

### CODE REVIEW

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java [hunk 1]
```diff
@@ -265,11 +265,11 @@ public InternalRange(StreamInput in) throws IOException {
         for (int i = 0; i < size; i++) {
             String key = in.getVersion().onOrAfter(Version.V_7_17_1) ? in.readOptionalString() : in.readString();
             double from = in.readDouble();
-            if (in.getVersion().onOrAfter(Version.V_7_17_0)) {
+            if (in.getVersion().onOrAfter(Version.V_7_17_0) && in.getVersion().before(Version.V_8_2_0)) {
                 in.readOptionalDouble();
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:269]: As a result of #83175 here we need to overwrite both `from` and `to` with whatever value we read from the wire if version matches `7.17.0`. Version `7.17.0` indeed, writes `from`, `originalFrom`, `to`, `originalTo` on the wire. In that case, if we just drop `originalFrom` and `originalTo` we have incorrect results because we would use the `from` and `to` during serialisation, which have precision issues.

**Review by nik9000 — APPROVED** [review]
LGTM though we should wait until #83715 is in and settled.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalBinaryRange.java [hunk 2]
```diff
@@ -70,7 +70,9 @@ private static String generateKey(BytesRef from, BytesRef to, DocValueFormat for
         }
 
         private static Bucket createFromStream(StreamInput in, DocValueFormat format, boolean keyed) throws IOException {
-            String key = in.getVersion().onOrAfter(Version.V_7_17_1) ? in.readOptionalString() : in.readString();
+            String key = in.getVersion().equals(Version.V_8_0_0) ? in.readString()
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalBinaryRange.java:73]: Maybe good to put a comment here in word saying what this does.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalBinaryRange.java [hunk 3]
```diff
@@ -81,10 +83,14 @@ private static Bucket createFromStream(StreamInput in, DocValueFormat format, bo
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
-            if (out.getVersion().onOrAfter(Version.V_7_17_1)) {
-                out.writeOptionalString(key);
-            } else {
+            if (out.getVersion().equals(Version.V_8_0_0)) {
                 out.writeString(key == null ? generateKey(from, to, format) : key);
+            } else {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalBinaryRange.java:88]: Maybe more readable with `} else if {`

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java [hunk 4]
```diff
@@ -157,17 +157,21 @@ private static String generateKey(double from, double to, DocValueFormat format)
...
+            } else {
+                if (out.getVersion().onOrAfter(Version.V_7_17_1)) {
+                    out.writeOptionalString(key);
+                } else {
+                    out.writeString(key == null ? generateKey(from, to, format) : key);
+                }
             }
             out.writeDouble(from);
-            if (out.getVersion().onOrAfter(Version.V_7_17_0)) {
+            if (out.getVersion().onOrAfter(Version.V_7_17_0) && out.getVersion().before(Version.V_8_2_0)) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:170]: I think you should move the v_8_2_0 thing to a different PR. it'll save a byte which is great but it makes the PR hard to backport.
